### PR TITLE
refactor(ci): Add molecule test into ovs_prepare ansible role

### PR DESCRIPTION
**Pull request by electrocucaracha**
## Summary

[Ansible Molecule](https://molecule.readthedocs.io/en/latest/) project provides a testing framework for Ansible playbooks helping to improve the quality and coverage of the code. This change improves the reproducibility and testing for `ovs_prepare` role.

## Test Plan

This change requires the [tox tool](https://pypi.org/project/tox/) which it's going to facilitate the installation of additional python dependencies on a virtual environment

There are some additional binary dependencies which can manually installed :
- [Vagrant](https://www.vagrantup.com/downloads)
- Libvirt or VirtualBox provisioners

Once those dependencies are installed, the `tox -e ovs_prepare` command will validate the role's execution.

## Additional Information

This change includes a GitHub action which executes this test on the CI.

### CODE REVIEW

File: lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml
Comment (pshelar) [inline lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml:22]: I am not sure why do we need changes to the ansible role tasks. Can we separate out changed to task and unit tests?

**Review by pshelar — CHANGES_REQUESTED** [review]

File: lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml
Comment (electrocucaracha) [reply lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml:22]: Those changes are needed to pass the Molecule idempotence and lint checks. maybe I can disable them to reduce the number of changes.

Comment (tmdzk) [reply lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml:22]: Can we disable them for now? I think our whole ansibles roles need a complete restructure. 
While we restructure them we'll be able to tackle the indempotence and lint changes.

**Review by tmdzk — CHANGES_REQUESTED** [review]
This PR has been in pending for a long time because some codeowners (myself inlcuded) forgot to review it. 
@electrocucaracha  thanks a lot for the contribution! Can we start small with the introduction of molecule and only execute verify for now? 
On top of that can we create a CI job that triggers molecule? (on another PR)

Thanks again for the contribution

**Comment by electrocucaracha** [comment]
> This PR has been in pending for a long time because some codeowners (myself inlcuded) forgot to review it.
> @electrocucaracha thanks a lot for the contribution! Can we start small with the introduction of molecule and only execute verify for now?
> On top of that can we create a CI job that triggers molecule? (on another PR)
> 
> Thanks again for the contribution

I can try to break down this in few commits to facilitate the code review

File: lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml
Comment (electrocucaracha) [reply lte/gateway/deploy/roles/ovs_prepare/tasks/main.yml:22]: I'm not sure if it's possible to disable them. I can investigate more about it.

**Comment by electrocucaracha** [comment]
@tmdzk I changed this PR in multiple commits and included a GH action to cover the molecule test execution. Hopefully that helps to clarify some dependencies and order of execution.

**Comment by ekowtaylor** [comment]
@electrocucaracha any plans to move this forward?

**Comment by electrocucaracha** [comment]
> @electrocucaracha any plans to move this forward?

I can rebase these commits., more likely there are changes which are not applicable now

**Comment by ekowtaylor** [comment]
I would say rebase, @tmdzk what are your thoughts about moving forward with this?

**Comment by electrocucaracha** [comment]
@ekowtaylor I have just rebased the changes and fix some tasks to be able to pass the molecule test. I noticed that [Jenkins is failing](https://jenkins.magmacore.org/job/magma_integration_tests/6884/console) given that it's looking for a magma_dev box  but the [last boxes](https://app.vagrantup.com/magmacore/boxes/magma_dev) were only created for VirtualBox provider

**Comment by ekowtaylor** [comment]
@tmdzk any thoughts on resolving the issues cited by @electrocucaracha ?

**Comment by ekowtaylor** [comment]
@tmdzk , @electronjoe could use your help to drive closure on this.

**Comment by electronjoe** [comment]
@ekowtaylor would it be possible to schedule a video chat with @electrocucaracha, @pshelar, @tmdzk if he's available and myself? Ideally we do not move on to adding new reviewers (e.g. myself) to a PR that has already had interaction with other reviewers (further, I'm ansible-ignorant).

I will comment on the Github workflow now, but I'm not a CI CODOWNER either so really I cannot help here.

File: .github/workflows/molecule_tests.yml
Comment (electronjoe) [inline .github/workflows/molecule_tests.yml:1]: Magma's github workflows don't have a copyright notice, but probably should.  If you are going to put one up here, lets use the template from other files below (Copyright 2021 The Magma Authors...).

File: .github/workflows/molecule_tests.yml
Comment (electronjoe) [inline .github/workflows/molecule_tests.yml:13]: I think just `pull_request` is sufficient - we've generally not been repeating tests on `push`.

File: .github/workflows/molecule_tests.yml
Comment (electronjoe) [inline .github/workflows/molecule_tests.yml:16]: Is it possible to limit the running of this GH Action to just those PRs that impact its success? Or is that **everything**?

If for example it is primarily validating ansible recipes, then maybe filtering only on changes to `/lte/gateway/deploy`? We now use a GH action for path filtering - if possible.  First you declare a `path_filter` step like [this one](https://github.com/magma/magma/blob/master/.github/workflows/gcc-problems.yml#L22-L32).  Then you add a `needs:` and `if:` clause to your `check-molecule` [like this](https://github.com/magma/magma/blob/master/.github/workflows/gcc-problems.yml#L35-L37).

This drastically accelerates CI results for most PRs that don't impact the test - and enables setting this workflow as `required` in Github (you cannot mark `required` if you use Github's native path based filtering, sadly).

File: .github/workflows/molecule_tests.yml
Comment (electronjoe) [inline .github/workflows/molecule_tests.yml:24]: Please drop the use of `actions/cache@v2`.  It was smart of you to cache, but unfortunately our project has a limit of 5GB of cache for GH Workflows and Vagrant / Docker builds will easily swallow that up and then some.  We are reserving the limited cache space for e.g. build tooling such as C++ caching (smaller needs, high impact on average PR).

**Review by electronjoe — COMMENTED** [review]
More tests are excellent =D

I'm ready to approve the GH Workflow subject to the requested changes - but I cannot speak for the broad Ansible changes. Further I'm not on CI CODEOWNERS yet so my LGTM does not count for much w.r.t. merge of `.github/**` files.

I think we had a miscommunication as well - in that I think Tim was suggesting that the PR efforts be split into several PRs so that each could be reviewed separately with increasing functionality.  That prevents what is happening here where the number of parties that have to LGTM grows and we start seeing long round trips.  But I get (in retrospect) that Tim's ask was ambiguous and interpreted as "please break this into commits".

**Comment by ekowtaylor** [comment]
> @ekowtaylor would it be possible to schedule a video chat with @electrocucaracha, @pshelar, @tmdzk if he's available and myself? Ideally we do not move on to adding new reviewers (e.g. myself) to a PR that has already had interaction with other reviewers (further, I'm ansible-ignorant).
> 
> I will comment on the Github workflow now, but I'm not a CI CODOWNER either so really I cannot help here.

@electrocucaracha, what's your availability like?

**Comment by electrocucaracha** [comment]
> > @ekowtaylor would it be possible to schedule a video chat with @electrocucaracha, @pshelar, @tmdzk if he's available and myself? Ideally we do not move on to adding new reviewers (e.g. myself) to a PR that has already had interaction with other reviewers (further, I'm ansible-ignorant).
> > I will comment on the Github workflow now, but I'm not a CI CODOWNER either so really I cannot help here.
> 
> @electrocucaracha, what's your availability like?

What about next week? Usually I have availability during my afternoons (2-5pm PST)

**Comment by ekowtaylor** [comment]
@electrocucaracha, I was hoping we can pick this up again and give it one last push to get it merged. Thoughts?

**Comment by electrocucaracha** [comment]
@ekowtaylor the last time that I tried many changes were not applicable anymore so I'm not sure if make sense to keep this PR open or start a new one with different approach

**Comment by ekowtaylor** [comment]
That's 

> @ekowtaylor the last time that I tried many changes were not applicable anymore so I'm not sure if make sense to keep this PR open or start a new one with different approach

That's a fair assessment. So how would you like to proceed?

**Comment by electrocucaracha** [comment]
You can close this PR for now and I'll try to create another new PR with changes that applies to the current state of the code.

**Comment by ekowtaylor** [comment]
Thanks so much, and please tag me once you have the new PR out for review.

**Pull request closed without merging**
